tlv/stream.go

```
@@ -0,0 +1,259 @@
+	var min Type
+	var overflow bool
+	for _, record := range records {
+		if overflow || record.typ < min {
+			panic(ErrStreamNotCanonical.Error())
+		}
+		if record.encoder == nil {
+			panic(ErrNoEncoder.Error())
+		}
+		if record.decoder == nil {
+			panic(ErrNoDecoder.Error())
+		}
+		if record.typ == math.MaxUint64 {
+			overflow = true
+		}
+		min = record.typ + 1
+	}
+
+	return &Stream{
+		records: records,
+	}
+}
+
+// Encode writes a Stream to the passed io.Writer. Each of the Records known to
+// the Stream is written in ascending order of their type so as to be canonical.
+//
+// The stream is constructed by concatenating the individual, serialized Records
+// where each record has the following format:
+//    [varint: type]
+//    [varint: length]
+//    [length: value]
+//
+// An error is returned if the io.Writer fails to accept bytes from the
+// encoding, and nothing else. The ordering of the Records is asserted upon the
+// creation of a Stream, and thus the output will be by definition canonical.
+func (s *Stream) Encode(w io.Writer) error {
+	// Iterate through all known records, if any, serializing each record's
+	// type, length and value.
+	for i := range s.records {
+		rec := &s.records[i]
+
+		// Write the record's type as a varint.
+		err := WriteVarInt(w, uint64(rec.typ), &s.buf)
+		if err != nil {
+			return err
+		}
+
+		// Write the record's length as a varint.
+		err = WriteVarInt(w, rec.Size(), &s.buf)
+		if err != nil {
+			return err
+		}
+
+		// Encode the current record's value using the stream's codec.
+		err = rec.encoder(w, rec.value, &s.buf)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// Decode deserializes TLV Stream from the passed io.Reader. The Stream will
+// inspect each record that is parsed and check to see if it has a corresponding
+// Record to facilitate deserialization of that field. If the record is unknown,
+// the Stream will discard the record's bytes and proceed to the subsequent
+// record.
```

> If the record is unknown,
> // the Stream will discard the record's bytes and proceed to the subsequent
> // record.

Is this subject to the even/odd rule? Also perhaps we should allow a "strict" stream that'll always error our rather than silently go to the next record to be decoded.

---

yes a strict mode would definitely be useful. the original version of this pr included that, but i removed it for now to simplify the diff